### DISCUSSION TITLE
Could/Should fauxmoESP be expanded to emulate other types of smart devices?  Request for comments...

### DISCUSSION THREAD

**Discussion by JonS9999** [vintlabs/fauxmoESP#151] [Ideas]
Hi all.

I'm hoping this isn't too far off the normal discussions of fauxmoESP (if it is, please don't shoot me ;-)

I'm enjoying fauxmo, and I *really* appreciate all the effort everyone has put into making it work.  Kudos to you all.

Here's my dilemma...  With all the great sales that places have been having with smart devices, I've been expanding out my home automation system.  I now have (real) Feit devices, Sonoff devices, Gosund devices, etc...  The problem is that each one of these types of devices requires their own app/Alexa skill in order to configure and control the devices.  This in itself is not a big deal, but some apps have more features than others.   I especially like the Sonoff app, as I was able to setup some nice scenes (rules) to have a 3-gang wall switch control some remote smart plugs and vise versa.  

A downside to using the off-the-shelf devices is that none of them support MQTT, which I would like in order to have these devices be controlled by (or inform) my Raspberry Pi/NodeRED home automation setup.

This got me thinking...  I wondering if it would make sense (or be possible) to expand fauxmoESP to be able to emulate different types of smart devices, such as the off-the-shelf Sonoff devices.  For example, if it would be possible to emulate a Sonoff 3-gang switch, then I could load up fauxmoESP (fauxSonOffESP?) onto a ESP8266/ESP32, link that ESP to my existing Sonoff app rules, then it could be a bridge between my RPi setup (via MQTT) and the real Sonoff devices and thus be controlled be both Alexa as well as MQTT.

Thoughts?

-Jon

**Comment by pvint** [comment]
Hi Jon,

Honestly, I don't have much recent experience with _real_ smart home devices! And, perhaps more surprisingly, I have little experience with MQTT either!

As far as emulating Sonoff goes - I would only be guessing as to how difficult that might be. At this time I feel my time is better spent on improving/refining what we are already doing, however if someone wants to do some research on how it works, I wouldn't complain! ;)

As for MQTT - as I mentioned, I'm not experience with it, but you should be able to use an MQTT library alongside FauxmoESP, and have it do `setState()` to control the fauxmo device? (Just a thought...)

This also reminds me that for reasons I don't recall, the `getState()` function was removed some time ago... should look into fixing that. 

Paul

**Reply by JonS9999** [reply]
Hi Paul.

Thanks for the quick reply.  Yeah, I'm just floating the idea out there -- I'm not proposing that you make the changes.  I'm on the "other side of the fence" where I haven't looked closely at the fauxmoESP code to know if it's structured to support just the Wemo devices or if it's generic enough (if those are the right words) to be able to expand it to emulate other smart devices.

In my copious amounts of free time (ha!), I will start trying to peel back the layers of the Sonoff devices to see what it would take to emulate a Sonoff switch or plug.  If I make any progress with that, then I'll look under the hook of the fauxmoESP code to see if it makes sense to incorporate the Sonoff code or if it makes more sense to setup a completely different project.  Right now it's too early to make those sorts of decisions.  :-)

Yeah, I would think that MQTT would run happily along side of fauxmoESP, however, in my setup I need to control non-fauxmoESP devices, specifically my Sonoff devices.  (Maybe I should shoot an email off to ITEAD (the makers of the Sonoff stuff) and see if they have any plans to incorporate MQTT into their off-the-shelf devices...)

Can fauxmoESP act as a trigger (a switch, for example)?  If so, then I might be able to create an Alexa routine (rule) that would say "When the fauxmoESP device ABC turns on, run the routine that sends a 'Light on' to the XYZ smart device".

Hey, take care, stay safe, don't stress and have a nice day.  :-)

-Jon

**Reply by WBorsato** [reply]
Hi @pvint  and Jon,

Do you have any progress on simulating other kind of devices? I'm playing with nodemcu +fauxmoESP + echo dot 4th gen, which shows my device as a philips lamp, but I wanted it to show as a simple smart switch. I have tried to change some parameters in template.h file but I got no success. Echo dot stopped finding my nodemcu.

Do you have any tips how to change your code from philips lamp to a smart switch?

@JonS9999, Maybe you could use pubsub library to your solution.

**Comment by pvint** [comment]
>  I haven't looked closely at the fauxmoESP code to know if it's structured to support just the Wemo devices

Surprise! It doesn't emulate Wemo, but rather it emulates Philips Hue!  (It originally was Wemo, but a few years back it got switched to Philips)

You could definitely use a Fauxmo device as a trigger in the code, something like:

```
if (action == "ON")
    SendMQTTMessage("blah");
```

No stress - you have a good day too, Jon!

**Reply by JonS9999** [reply]
Right-o.  I'm wondering if I can go the other way -- instead of Alexa sending a "turn on" command to the fauxmoESP device, have a button on the fauxmoESP device such that when the button is pressed, the fauxmoESP device sends a "I was just turned on" message to Alexa (to trigger an Alexa routine).   Hmmmmm...

It looks like the latest-and-greatest fauxmoESP code is in the official Arduino library archive/repo, so I'll upgrade my library (from 3.1.2 to 3.2) and start peeking at the code.

Cheers.

-Jon
